c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -699,10 +854,27 @@ class DeviceCachingAllocator {
     // Free all non-split cached blocks
     free_blocks(large_blocks);
     free_blocks(small_blocks);
+
+    for (auto it = graph_pools.begin(); it != graph_pools.end(); ) {
```

Comment:
When `use_count == 0`, is it possible for the use count to go back up again? If not, a simple way to do cleanup but maintain the correct asymptotics when there are many live mempools is to just move mempools to a free list when their use count goes to zero, so you only have to iterate over the free list at this point.

Reply:
> When use_count == 0, is it possible for the use count to go back up again

hmm definitely not intended, but I need to think if it could...

> just move mempools to a free list when their use count goes to zero, so you only have to iterate over the free list at this point.

that's a better idea, I'll make that change.

Reply:
Created secondary map (`graph_pools_freeable`) to avoid iterating through all `graph_pools` in free_cached_blocks.